dali/driver/hasseb.py

```diff
@@ -34,160 +50,410 @@ def __repr__(self):
-
-    This code borrows from the HassebDALIUSBDriver.
     """
+    device_found = None
     logger = logging.getLogger('HassebDALIUSBDriver')
+    sn = 0
+    send_message = None
+    _pending = None
+    _response_message = None
+
+    def __init__(self):
+        try:
+            self.device = hidapi.hid_open(HASSEB_USB_VENDOR, HASSEB_USB_PRODUCT, None)
+            self.device_found = 1
+        except:
+            self.device_found = None
+
+    def wait_for_response(self):
+        raise NotImplementedError()
 
     def construct(self, command):
+        # sequence number
+        self.sn = self.sn+1
+        if self.sn > 255:
+            self.sn = 1
+        frame_length = 16
+        if command.is_query:
+            expect_reply = 1
+        else:
+            expect_reply = 0
+        transmitter_settling_time = 0
+        if command.is_config:
+            send_twice = 10 # 10 ms delay between messages
+        else:
+            send_twice = 0
         frame = command.frame.as_byte_sequence
         byte_a, byte_b = frame
-        data = struct.pack('BB', byte_a, byte_b)
-
+        data = struct.pack('BBBBBBBBBB', 0xAA, HASSEB_DALI_FRAME, self.sn,
+                           frame_length, expect_reply,
+                           transmitter_settling_time, send_twice,
+                           byte_a, byte_b,
+                           0)
         return data
 
     def extract(self, data):
-        if len(data) >= 2:
-            response_status = data[0]
-            if response_status == HASSEB_DRIVER_NO_DATA_AVAILABLE:
-                # 0: "No Data Available"
-                self.logger.debug("No Data Available")
-                return HassebDALIUSBNoDataAvailable()
-            elif response_status == HASSEB_DRIVER_NO_ANSWER:
+        if data == None:
+            return None
+        elif data[1] == HASSEB_DRIVER_NO_DATA_AVAILABLE:
+            # 0: "No Data Available"
+            self.logger.debug("No Data Available")
+            return HassebDALIUSBNoDataAvailable()
+        elif data[1] == HASSEB_DALI_FRAME:
+            response_status = data[3]
+            if response_status == HASSEB_DRIVER_NO_ANSWER:
                 # 1: "No Answer"
                 self.logger.debug("No Answer")
                 return HassebDALIUSBNoAnswer()
-            elif response_status == HASSEB_DRIVER_OK:
+            elif response_status == HASSEB_DRIVER_OK and data[4] == 1:
                 # 2: "OK"
-                return BackwardFrame(data[1])
+                return BackwardFrame(data[5])
             elif response_status == HASSEB_DRIVER_INVALID_ANSWER:
                 # 3: "Invalid Answer"
                 return BackwardFrameError(255)
+            elif response_status == HASSEB_DRIVER_TOO_EARLY:
+                # 4: "Answer too early"
+                self.logger.debug("Answer too early")
+                return HassebDALIUSBAnswerTooEarly()
+            elif response_status == HASSEB_DRIVER_SNIFFER_BYTE:
```

Comment:
It looks like this ought to be an exception, rather than a 'magic' return value.